ledger/object/record_db.go

```diff
@@ -229,29 +258,92 @@ func setPosition(txn *badger.Txn, recID insolar.ID, position uint32) error {
 	return txn.Set(fullKey, recID.Bytes())
 }
 
-// TruncateHead remove all records after lastPulse
-func (r *RecordDB) TruncateHead(ctx context.Context, from insolar.PulseNumber) error {
-	it := store.NewReadIterator(r.db.Backend(), recordKey(*insolar.NewID(from, nil)), false)
+func (r *RecordDB) truncateRecordsHead(ctx context.Context, from printableKey, km keyMaker) error {
+	it := store.NewReadIterator(r.db.Backend(), from, false)
+	defer it.Close()
+
+	var hasKeys bool
+	for it.Next() {
+		hasKeys = true
+		key := km(it.Key())
+
+		err := r.db.Delete(key)
+		if err != nil {
+			return errors.Wrapf(err, "can't delete key: %s", key.String())
+		}
+
+		inslogger.FromContext(ctx).Debugf("Erased key: %s", key.String())
+	}
+
+	if !hasKeys {
+		inslogger.FromContext(ctx).Infof("No records. Nothing done. Start key: %s", from.String())
+	}
+
+	return nil
+}
+
+func (r *RecordDB) truncatePositionRecordHead(ctx context.Context, from printableKey, km keyMaker, prefix byte) error {
+	it := store.NewReadIterator(r.db.Backend(), from, false)
 	defer it.Close()
 
 	var hasKeys bool
 	for it.Next() {
 		hasKeys = true
-		key := newRecordKey(it.Key())
-		keyID := insolar.ID(key)
+		if it.Key()[0] != prefix {
+			continue
+		}
+		key := km(it.Key())
 
-		err := r.db.Backend().Update(func(txn *badger.Txn) error {
-			fullKey := append(key.Scope().Bytes(), key.ID()...)
-			return txn.Delete(fullKey)
-		})
+		err := r.db.Delete(key)
 		if err != nil {
-			return errors.Wrapf(err, "can't delete key: %+v", key)
+			return errors.Wrapf(err, "can't delete key: %s", key.String())
 		}
-		inslogger.FromContext(ctx).Debugf("Erased key with pulse number: %s. ID: %s", keyID.Pulse().String(), keyID.String())
+
+		inslogger.FromContext(ctx).Debugf("Erased key: %s", key.String())
 	}
 
 	if !hasKeys {
-		inslogger.FromContext(ctx).Infof("No records. Nothing done. Pulse number: %s", from.String())
+		inslogger.FromContext(ctx).Infof("No records. Nothing done. Start key: %s", from.String())
+	}
+
+	return nil
+}
+
+type printableKey interface {
+	store.Key
+	String() string
+}
+
+type keyMaker func([]byte) printableKey
+
+func recordKeyMaker(raw []byte) printableKey {
+	return newRecordKey(raw)
+}
+
+func positionKeyMaker(raw []byte) printableKey {
+	switch raw[0] {
+	case recordPositionKeyPrefix:
+		return newRecordPositionKeyFromBytes(raw)
+	case lastKnownRecordPositionKeyPrefix:
+		return newLastKnownRecordPositionKey(raw)
+	default:
+		panic("unknown prefix: " + string(raw[0]))
+	}
+}
+
+// TruncateHead remove all records after lastPulse
+func (r *RecordDB) TruncateHead(ctx context.Context, from insolar.PulseNumber) error {
+
+	if err := r.truncateRecordsHead(ctx, recordKey(*insolar.NewID(from, nil)), recordKeyMaker); err != nil {
+		return errors.Wrap(err, "failed to truncate records head")
+	}
+
+	if err := r.truncatePositionRecordHead(ctx, recordPositionKey{pn: from}, positionKeyMaker, recordPositionKeyPrefix); err != nil {
```

Comment:
You can either create two different functions for that or make an anonymus function with parameter inside. Both solutions are cleaner and simpler to read.

Reply:
They will be absolutely the same. It's copy-past. 
What is the advantage of anonymous function here?